compiler/rustc_codegen_ssa/src/target_features.rs

```diff
@@ -62,39 +65,74 @@ pub fn from_target_feature(
+                    rust_features.avx512_target_feature
+                }
+                Stability::Unstable(sym::sse4a_target_feature) => {
+                    rust_features.sse4a_target_feature
+                }
+                Stability::Unstable(sym::tbm_target_feature) => rust_features.tbm_target_feature,
+                Stability::Unstable(sym::wasm_target_feature) => rust_features.wasm_target_feature,
+                Stability::Unstable(sym::rtm_target_feature) => rust_features.rtm_target_feature,
+                Stability::Unstable(sym::ermsb_target_feature) => {
+                    rust_features.ermsb_target_feature
+                }
+                Stability::Unstable(sym::bpf_target_feature) => rust_features.bpf_target_feature,
+                Stability::Unstable(sym::aarch64_ver_target_feature) => {
+                    rust_features.aarch64_ver_target_feature
+                }
+                Stability::Unstable(sym::csky_target_feature) => rust_features.csky_target_feature,
+                Stability::Unstable(sym::loongarch_target_feature) => {
+                    rust_features.loongarch_target_feature
+                }
+                Stability::Unstable(sym::lahfsahf_target_feature) => {
+                    rust_features.lahfsahf_target_feature
+                }
+                Stability::Unstable(sym::prfchw_target_feature) => {
+                    rust_features.prfchw_target_feature
+                }
+                Stability::Unstable(sym::sha512_sm_x86) => rust_features.sha512_sm_x86,
+                Stability::Unstable(sym::x86_amx_intrinsics) => rust_features.x86_amx_intrinsics,
+                Stability::Unstable(sym::xop_target_feature) => rust_features.xop_target_feature,
+                Stability::Unstable(sym::s390x_target_feature) => {
+                    rust_features.s390x_target_feature
+                }
+                Stability::Unstable(name) => bug!("unknown target feature gate {}", name),
             };
             if !allowed {
-                feature_err(
-                    &tcx.sess,
-                    feature_gate.unwrap(),
-                    item.span(),
-                    format!("the target feature `{feature}` is currently unstable"),
-                )
-                .emit();
+                match stability {
+                    Stability::Stable => unreachable!(),
+                    &Stability::Unstable(lang_feature_name) => {
+                        feature_err(
+                            &tcx.sess,
+                            lang_feature_name,
+                            item.span(),
+                            format!("the target feature `{feature}` is currently unstable"),
+                        )
+                        .emit();
+                    }
+                    Stability::Forbidden { reason } => {
+                        tcx.dcx().emit_err(errors::ForbiddenTargetFeature {
```

Comment:
This code here is only used for `#[target_features]` (the attribute, not the `-C` flag), where unknown features are a hard error. Thus forbidden features should also be immediately a hard error here, this is not a breaking change.